src/Streamly/Internal/Data/Parser/ParserD.hs

```diff
@@ -817,45 +817,53 @@ spanByRolling eq f1 f2 =
 
 -- | See 'Streamly.Internal.Data.Parser.takeP'.
 --
+-- /Internal/
 {-# INLINE takeP #-}
 takeP :: Monad m => Int -> Parser m a b -> Parser m a b
-takeP lim (Parser fstep finitial fextract) = Parser step initial extract
+takeP lim (Parser pstep pinitial pextract) = Parser step initial extract
 
     where
 
     initial = do
-        res <- finitial
-        return
-            $ case res of
-                  IPartial s -> IPartial (Tuple' 0 s)
-                  IDone b -> IDone b
-                  IError e -> IError e
-
-    step (Tuple' cnt r) a
-        | cnt < lim = do
-            res <- fstep r a
+            res <- pinitial
+            case res of
+                IPartial s ->
+                    if lim > 0
+                    then return $ IPartial $ Tuple' 0 s
+                    else IDone <$> pextract s
+                IDone b -> return $ IDone b
+                IError e -> return $ IError e
+
+    step (Tuple' cnt r) a = do
+            assert (cnt < lim) (return ())
+            res <- pstep r a
             let cnt1 = cnt + 1
             case res of
+                Partial 0 s -> do
+                    assert (cnt1 >= 0) (return ())
+                    if cnt1 < lim
+                    then return $ Partial 0 $ Tuple' cnt1 s
+                    else Done 0 <$> pextract s
+                Continue 0 s -> do
+                    assert (cnt1 >= 0) (return ())
+                    if cnt1 < lim
+                    then return $ Continue 0 $ Tuple' cnt1 s
+                    -- XXX This should error out?
+                    -- If designed properly, this will probably error out.
+                    -- "pextract" should error out
```

Comment:
This is a tricky case, we have the following options:

1. `Done 0` with extract as you have written
2. `Done n`, will require buffering elements
3. Use a backtracking fold and not a parser, once we have backtracking in folds